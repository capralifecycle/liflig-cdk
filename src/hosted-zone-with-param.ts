import * as constructs from "constructs"
import * as route53 from "aws-cdk-lib/aws-route53"
import * as ssm from "aws-cdk-lib/aws-ssm"
import * as cdk from "aws-cdk-lib"
import { SsmParameterReader } from "./ssm-parameter-reader"
import { getStageOrApp } from "./utils"

interface Props extends route53.HostedZoneProps {
  /**
   * We don't expect a hosted zone to be recreated while it is being
   * referenced in other stacks, but in case it is, this value can be
   * used to force consumers to refresh the value.
   *
   * @default 1
   */
  nonce?: string
}

/**
 * A Hosted Zone that writes its ID to SSM Parameter Store and provides
 * a helper to easily retrieve the Hosted Zone cross-region.
 */
export class HostedZoneWithParam extends constructs.Construct {
  private readonly nonce: string
  private readonly hostedZone: route53.HostedZone
  readonly name: string
  readonly idParamName: string

  constructor(scope: constructs.Construct, id: string, props: Props) {
    super(scope, id)

    this.nonce = props.nonce ?? "1"
    this.name = props.zoneName
    this.idParamName = `/cf/hosted-zone-id/${props.zoneName}-${
      props.vpcs ? "private" : "public"
    }`

    this.hostedZone = new route53.HostedZone(this, "Resource", props)

    new ssm.CfnParameter(this, "IdParam", {
      type: ssm.ParameterType.STRING,
      name: this.idParamName,
      value: this.hostedZone.hostedZoneId,
    })

    if (this.hostedZone.hostedZoneNameServers) {
      new cdk.CfnOutput(this, "Nameservers", {
        value: cdk.Fn.join(", ", this.hostedZone.hostedZoneNameServers),
      })
    }

    new cdk.CfnOutput(this, "Id", {
      value: this.hostedZone.hostedZoneId,
    })
  }

  /**
   * Get the Hosted Zone by resolving the zone id from SSM Parameter Store
   * in case we are cross-region.
   */
  public getHostedZone(
    scope: constructs.Construct,
    id: string,
  ): route53.IHostedZone {
    const hostedZoneRegion = cdk.Stack.of(this).region
    const consumerRegion = cdk.Stack.of(scope).region

    const sameStageOrApp = getStageOrApp(this) === getStageOrApp(scope)

    // Fast-path: Same region and parent stage/app.
    if (hostedZoneRegion === consumerRegion && sameStageOrApp) {
      return this.hostedZone
    }

    // Only add dependency if within same app/stage. If not it
    // is the caller responsibility to ensure deployment order.
    if (sameStageOrApp) {
      scope.node.addDependency(this)
    }

    const hostedZoneId = new SsmParameterReader(scope, `${id}Param`, {
      parameterName: this.idParamName,
      region: cdk.Stack.of(this).region,
      nonce: this.nonce,
    }).getParameterValue()
    return route53.HostedZone.fromHostedZoneAttributes(scope, id, {
      hostedZoneId,
      zoneName: this.name,
    })
  }
}
