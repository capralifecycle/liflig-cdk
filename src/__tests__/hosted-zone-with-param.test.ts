import { App, CfnOutput, Stack } from "@aws-cdk/core"
import "jest-cdk-snapshot"
import { HostedZoneWithParam } from "../hosted-zone-with-param"

test("hosted-zone-with-param for same region", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1", {
    env: {
      region: "eu-west-1",
    },
  })
  const stack2 = new Stack(app, "Stack2", {
    env: {
      region: "eu-west-1",
    },
  })

  const hostedZone1 = new HostedZoneWithParam(stack1, "HostedZone", {
    zoneName: "example.com",
  })

  const hostedZone2 = hostedZone1.getHostedZone(stack2, "HostedZone")

  new CfnOutput(stack2, "HostedZoneId", {
    value: hostedZone2.hostedZoneId,
  })

  expect(stack1).toMatchCdkSnapshot()
  expect(stack2).toMatchCdkSnapshot()
})

test("hosted-zone-with-param for different region", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1", {
    env: {
      region: "eu-west-1",
    },
  })
  const stack2 = new Stack(app, "Stack2", {
    env: {
      region: "us-east-1",
    },
  })

  const hostedZone1 = new HostedZoneWithParam(stack1, "HostedZone", {
    zoneName: "example.com",
  })

  const hostedZone2 = hostedZone1.getHostedZone(stack2, "HostedZone")

  new CfnOutput(stack2, "HostedZoneId", {
    value: hostedZone2.hostedZoneId,
  })

  expect(stack1).toMatchCdkSnapshot()
  expect(stack2).toMatchCdkSnapshot()
})
