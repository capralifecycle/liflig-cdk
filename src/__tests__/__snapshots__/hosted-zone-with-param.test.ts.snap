// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`hosted-zone-with-param for different region 1`] = `
Object {
  "Outputs": Object {
    "HostedZoneIdFC741845": Object {
      "Value": Object {
        "Ref": "HostedZoneA9DD9B89",
      },
    },
    "HostedZoneNameservers0897D4B3": Object {
      "Value": Object {
        "Fn::Join": Array [
          ", ",
          Object {
            "Fn::GetAtt": Array [
              "HostedZoneA9DD9B89",
              "NameServers",
            ],
          },
        ],
      },
    },
  },
  "Resources": Object {
    "HostedZoneA9DD9B89": Object {
      "Properties": Object {
        "Name": "example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "HostedZoneIdParamE55EE91C": Object {
      "Properties": Object {
        "Name": "/cf/hosted-zone-id/example.com-public",
        "Type": "String",
        "Value": Object {
          "Ref": "HostedZoneA9DD9B89",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`hosted-zone-with-param for different region 2`] = `
Object {
  "Outputs": Object {
    "HostedZoneId": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "HostedZoneParam13DE3CEA",
          "Parameter.Value",
        ],
      },
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "34a66902956b031404ef497526f619b900363fe7fd65ff02b1de4c30fe10c034.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HostedZoneParam13DE3CEA": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "HostedZoneParamCustomResourcePolicyE355AB96",
      ],
      "Metadata": Object {
        "Nonce": "1",
      },
      "Properties": Object {
        "Create": "{\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/cf/hosted-zone-id/example.com-public\\"},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/cf/hosted-zone-id/example.com-public\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/cf/hosted-zone-id/example.com-public\\"},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/cf/hosted-zone-id/example.com-public\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "HostedZoneParamCustomResourcePolicyE355AB96": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/cf/hosted-zone-id/example.com-public",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HostedZoneParamCustomResourcePolicyE355AB96",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`hosted-zone-with-param for same region 1`] = `
Object {
  "Outputs": Object {
    "ExportsOutputRefHostedZoneA9DD9B896D7C4880": Object {
      "Export": Object {
        "Name": "Stack1:ExportsOutputRefHostedZoneA9DD9B896D7C4880",
      },
      "Value": Object {
        "Ref": "HostedZoneA9DD9B89",
      },
    },
    "HostedZoneIdFC741845": Object {
      "Value": Object {
        "Ref": "HostedZoneA9DD9B89",
      },
    },
    "HostedZoneNameservers0897D4B3": Object {
      "Value": Object {
        "Fn::Join": Array [
          ", ",
          Object {
            "Fn::GetAtt": Array [
              "HostedZoneA9DD9B89",
              "NameServers",
            ],
          },
        ],
      },
    },
  },
  "Resources": Object {
    "HostedZoneA9DD9B89": Object {
      "Properties": Object {
        "Name": "example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "HostedZoneIdParamE55EE91C": Object {
      "Properties": Object {
        "Name": "/cf/hosted-zone-id/example.com-public",
        "Type": "String",
        "Value": Object {
          "Ref": "HostedZoneA9DD9B89",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`hosted-zone-with-param for same region 2`] = `
Object {
  "Outputs": Object {
    "HostedZoneId": Object {
      "Value": Object {
        "Fn::ImportValue": "Stack1:ExportsOutputRefHostedZoneA9DD9B896D7C4880",
      },
    },
  },
}
`;
